Core/Entities: Warlock pet base damage table fix

PolarCookie opened the pull request:
remake of https://github.com/TrinityCore/TrinityCore/pull/19360 on a new branch
tldr: suggested fix to the warlock pet base damage, placing base damage into the pet_levelstats table.
Fills world.pet_levelstats with base damage values and stats for the 5 common warlock pets for all levels.

references:
[1] simcraft
https://storage.googleapis.com/google-code-archive-downloads/v2/code.google.com/simulationcraft/simc-335-1-win32.zip
[2] old wowwiki
http://wowwiki.wikia.com/wiki/Felguard_(warlock_minion)?oldid=2409171
http://wowwiki.wikia.com/wiki/Felhunter?oldid=2714647
http://wowwiki.wikia.com/wiki/Succubus_(warlock_minion)?oldid=2408937
http://wowwiki.wikia.com/wiki/Voidwalker_(warlock_minion)?oldid=2353207
http://wowwiki.wikia.com/wiki/Imp_(warlock_minion)?oldid=2410392
[3] voidwalker tanking
https://www.youtube.com/watch?v=SWqecvALz6Q
https://www.youtube.com/watch?v=WO-E2craHms
[4] my work files
https://www.mediafire.com/folder/rbt5u2yu1w232a9,7mt04maidc7u0uu,gv97znicm4y5sb5,rv7jsd0d32gc7sc,990y8fgqhk5vgcc,je7y4dfqg574cug,quft16relsflk6f,ur88x7lih6d56x1/shared

![base_dmg_nonmono](https://cloud.githubusercontent.com/assets/1158801/24726724/a37d1aaa-1a53-11e7-919b-5eccf267a603.png)
This is a plot of all the wowwiki datapoints, with stock trinity and sunwells felguard for comparison. There are many datapoints that are missing, if anyone have any sources to fill in let me know asap, but there's reason to suspect Blizzard reduced the damage at lvl 60 and 70 for balancing reason back in the day. Other than that it seems to be more or less smooth curves.

Dalaran-wow and ToD (rip) solved this issue by linear scaling; dalaranwow/dalaran-wow#3237.
This works for lvl 80, but all pets become massively overpowered at lower levels.

![base_dmg](https://cloud.githubusercontent.com/assets/1158801/24724040/ff8fbdf2-1a49-11e7-93d0-f3bccd42b791.png)
This solution will scale the pets as illustrated on this image.
I interpolated the tables from wowwiki and kept it strictly monotonic, ignoring damage values lower than previous levels. This can be discussed, but the alternative would be the zigzagged progression as shown on the graph above. I suspect this monotonic will be closer to the actual blizzlike values as the wowwiki table is missing far to many data points in the 60-80 range.
The sql file for non-monotonic damage values can be generated by the python scripts in my work files. Edit the calls to _interpolate_ in pet_stats.py to have False as argument and run it.


Target branch(es): 3.3.5

- [x] 3.3.5

PolarCookie committed:
fixed pet base damage and stats for the 5 common warlock pets for all levels, added base damage to the pet_levelstats table

PolarCookie committed:
fixed pet base damage and stats for the 5 common warlock pets for all levels, added base damage to the pet_levelstats table

Commit:
Update ObjectMgr.cpp

fixed tab/space issue

PolarCookie commented:
looking at the interpolated graphs, the weird shape of voidwalker really sticks out.
Quote from his wowwiki page 
> Voidwalker stats have severely changed in some areas, so the list needs most levels to be updated with base pet stats).
Looking over the history, indeed some parts of his damage is from older wow version.
Does anyone know if he was buffed in wotlk?
I could remove the ranges 12-34 to smooth them out and run the interpolation again, 42-60 is ignored because by the monotonic setting.

Commit:
Update 9999_99_99_99_world.sql

updated voidwalker damage to pure wotlk values

Riztazz commented:
You need to drop that world. prefix in your sql, not everyone has their databases called that :P

PolarCookie commented:
good point, halfway through updating voidwalker.
sec

Commit:
Update 9999_99_99_99_world.sql

removed world.

PolarCookie commented:
removed world.
updated the mediafire files
fixed voidwalker: 
http://wowwiki.wikia.com/wiki/Patch_3.0.2
>  Voidwalker: Health increased by 20%, Damage increased by 16%. The amount of attack power the Voidwalker gains from its master's spell damage has been increased by 45%. 

new values are interpolated only from values posted after 3.0.2, only levels 12-34 are different.
looks like this now
![base_dmg](https://cloud.githubusercontent.com/assets/1158801/24753638/3d3419ee-1ad3-11e7-8d87-f2753a4dc1ed.png)

Aokromes commented on sql/updates/world/3.3.5/9999_99_99_99_world.sql:
```diff
@@ -88,29 +88,29 @@ UPDATE world.pet_levelstats SET str=297, agi=79, sta=118, inte=369, spi=367, arm
-UPDATE world.pet_levelstats SET str=62, agi=49, sta=131, inte=74, spi=54, armor=1812, min_dmg=21, max_dmg=26 WHERE creature_entry=1860 AND level=33;
-UPDATE world.pet_levelstats SET str=63, agi=50, sta=135, inte=76, spi=56, armor=1861, min_dmg=27, max_dmg=38 WHERE creature_entry=1860 AND level=34;
+UPDATE world.pet_levelstats SET str=31, agi=26, sta=36, inte=24, spi=31, armor=876, min_dmg=10, max_dmg=16 WHERE creature_entry=1860 AND level=12;
+UPDATE world.pet_levelstats SET str=32, agi=28, sta=43, inte=26, spi=32, armor=877, min_dmg=11, max_dmg=17 WHERE creature_entry=1860 AND level=13;
```
remove world. part add ` around table and database names.

ccrs commented:
Why dont u just delete and insert? IMO updating all the values of the table with new data is just worse.

PolarCookie commented:
most of the work was getting the right values, didn't give it much thought.
can edit to make it so if wanted.

PolarCookie commented:
no, can't do a delete/insert as then you'd lose the base hp/mana values.
I'd very much prefer to get some data about those and include them in my spreedsheet, but alas.

Aokromes committed:
Update 9999_99_99_99_world.sql

PolarCookie commented:
about the succubus
simcraft says she has the same base damage at lvl 80 as felguard; 412.5 damage, but she has a 1.05 damage_modifier, calculator says 433.125.
From the wowwiki table I got 462 base damage after subtracting AP bonus.
From the wowwiki also there's this quote.
> Level 80 stats taken from naked untalented Undead Warlock

That's all I know about that

Riztazz commented:
@PolarCookie 
Could you poke me on Trinitys IRC, i got a question, same name as here

Commit:
SQL codestyle

- added missing back ticks for column names + table names
- aligned columns for better overview of value variations

Commit:
Update ObjectMgr.h

armor is stored in and read from the database as int32

Commit:
Update ObjectMgr.cpp

min_dmg is stored in struct and database as int16, read it too as such

Keader commented:
Looks a nice work

Aokromes commented:
Conflicting files
src/server/game/Globals/ObjectMgr.h

Commit:
Merge branch '3.3.5' into warlock_pet_base_damage_table_fix

Commit:
rename

DanVS commented:
Any news on this?

Aokromes commented:
You tested it?

DanVS commented:
Yes (manually imported). Builds and works in-game.

tkrokli commented:
Thanks for the feedback, but did you also test if it solves any known warlock pet scaling issues in current TC branch?

DanVS commented:
I didn't check that it details, but pet stats and damage are much better (and closer to retail videos) then they were, with this PR.

tkrokli commented:
@PolarCookie : What is your take on this PR, is it ready or does it need any more work?

PolarCookie commented:
@tkrokli, I think it's fine as is for warlocks.

I'm not happy that you can't rerun the sql because of the add column; it will error out, but it's mostly an annoyance.

I wanted to get all the other pets into the new columns, to reduce the huge switch case, but never got around to it. I tried finding some data on what hunter pets were supposed to have without any luck. Can be done later.

jackpoz commented:
can this be merged ?

Aokromes commented:
Conflicting files
src/server/game/Globals/ObjectMgr.cpp

Commit:
Merge branch '3.3.5' into warlock_pet_base_damage_table_fix

Aokromes commented:
@jackpoz click shinny button xd

Aokromes committed:
Rename 9999_99_99_99_world.sql to 2018_04_15_00_world.sql
